python/ray/serve/_private/replica.py

```diff
@@ -814,145 +811,199 @@ async def call_reconfigure(self, user_config: Any):
+    def _prepare_args_for_grpc_request(
+        self,
+        request: gRPCRequest,
+        request_metadata: RequestMetadata,
+        user_method_params: Dict[str, inspect.Parameter],
+    ) -> Tuple[Tuple[Any], Dict[str, Any]]:
+        """Prepare arguments for a user method handling a gRPC request.
 
-        runner_method = None
-        try:
-            runner_method = self._get_user_callable_method(request_metadata.call_method)
-            if inspect.isgeneratorfunction(runner_method) or inspect.isasyncgenfunction(
-                runner_method
-            ):
-                raise TypeError(
-                    f"Method '{runner_method.__name__}' is a generator function. "
-                    "You must use `handle.options(stream=True)` to call "
-                    "generators on a deployment."
-                )
+        Returns (request_args, request_kwargs).
+        """
+        request_args = (pickle.loads(request.grpc_user_request),)
+        if GRPC_CONTEXT_ARG_NAME in user_method_params:
+            request_kwargs = {GRPC_CONTEXT_ARG_NAME: request_metadata.grpc_context}
+        else:
+            request_kwargs = {}
 
-            method_to_call = sync_to_async(runner_method)
+        return request_args, request_kwargs
 
-            # Edge case to support empty HTTP handlers: don't pass the Request
-            # argument if the callable has no parameters.
-            params = inspect.signature(runner_method).parameters
-            if request_metadata.is_http_request and len(params) == 0:
-                request_args, request_kwargs = tuple(), {}
-            elif request_metadata.is_grpc_request and GRPC_CONTEXT_ARG_NAME in params:
-                request_kwargs = {GRPC_CONTEXT_ARG_NAME: request_metadata.grpc_context}
-
-            result = await method_to_call(*request_args, **request_kwargs)
-            if inspect.isgenerator(result) or inspect.isasyncgen(result):
+    async def _handle_user_method_result(
+        self,
+        result: Any,
+        user_method_name: str,
+        request_metadata: RequestMetadata,
+        *,
+        result_queue: Optional[MessageQueue],
+        is_asgi_app: bool,
+        asgi_args: Optional[ASGIArgs],
+    ) -> Any:
+        """Postprocess the result of a user method.
+
+        If this is for a streaming request, the result should be a generator that will
+        be consumed and its outputs placed on the `result_queue`.
+
+        This method may raise an exception if the result is not of the expected type
+        (e.g., non-generator for streaming requests or generator for unary requests).
+        """
+        result_is_gen = inspect.isgenerator(result)
+        result_is_async_gen = inspect.isasyncgen(result)
+        if request_metadata.is_streaming:
+            if result_is_gen:
+                for r in result:
+                    if request_metadata.is_grpc_request:
+                        r = (request_metadata.grpc_context, r.SerializeToString())
```

Comment:
Why do we need to channel back `grpc_context`?

Reply:
this is also existing code just moved. but yes it's a bit ugly. we need to channel back the context because it contains user-provided information such as the status code and trailing metadata, but really we should be doing this only in the final message rather than in every single message (cc @GeneDer)

Reply:
But we're modifying object in-place, right? Once we deserialized context from the wire it's gonna be a single object that's gonna be mutated, so my point is we don't even need to pass it back we can just let the caller serialize it and send it back

Reply:
Yes, that's my point we should really only need to serialize and send it back once (can be at the caller too doesn't matter)

Reply:
We should also unify this as much as possible with the unary path. There's an ugly conditional a few lines of code below handling it right now

Reply:
I think we might still need to run do this at every iteration just in case the connect is dropped in between or if there are some exception raised which terminated the request early? (but could be a small edge case maybe it's not worth considering). Need to test a bit more on how it behave and created this issue to do some cleanups https://github.com/ray-project/ray/issues/42347

Btw, I think there is still a difference between the HTTP and gRPC results that we also need to call `SerializeToString`, but I guess that can be moved more downstream later as well.

Reply:
If connection is terminated, stream will be interrupted. We don't need to have special handling in that case